pkg/flow/adapter/synchronizer/adapter.go

```diff
@@ -0,0 +1,147 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package synchronizer
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	pkgadapter "knative.dev/eventing/pkg/adapter/v2"
+	"knative.dev/pkg/logging"
+)
+
+var _ pkgadapter.Adapter = (*adapter)(nil)
+
+type adapter struct {
+	ceClient cloudevents.Client
+	logger   *zap.SugaredLogger
+
+	correlationKey  *CloudEventKey
+	responseTimeout time.Duration
+
+	sessions *storage
+	sinkURL  string
+}
+
+// NewAdapter returns adapter implementation.
+func NewAdapter(ctx context.Context, envAcc pkgadapter.EnvConfigAccessor, ceClient cloudevents.Client) pkgadapter.Adapter {
+	env := envAcc.(*envAccessor)
+	logger := logging.FromContext(ctx)
+
+	timeout, err := time.ParseDuration(env.ResponseWaitTimeout)
+	if err != nil {
+		logger.Panic("Failed to parse timeout value: %v", err)
+	}
+
+	key, err := NewCorrelationKey(env.CorrelationKey, env.CorrelationKeyLength)
+	if err != nil {
+		logger.Panic("Cannot create an instance of Correlation Key: %v", err)
+	}
+
+	return &adapter{
+		ceClient:       ceClient,
+		logger:         logger,
+		correlationKey: key,
+
+		responseTimeout: timeout,
+		sessions:        newStorage(),
+		sinkURL:         env.Sink,
+	}
+}
+
+// Returns if stopCh is closed or Send() returns an error.
+func (a *adapter) Start(ctx context.Context) error {
+	a.logger.Info("Starting Synchronizer Adapter")
+	return a.ceClient.StartReceiver(ctx, a.dispatch)
+}
+
+func (a *adapter) dispatch(ctx context.Context, event cloudevents.Event) (*cloudevents.Event, cloudevents.Result) {
+	a.logger.Debugf("Received the event: %s", event.String())
+
+	if correlationID, exists := a.correlationKey.Get(event); exists {
+		return a.serveResponse(ctx, correlationID, event)
+	}
+
+	correlationID := a.correlationKey.Set(&event)
+	return a.serveRequest(ctx, correlationID, event)
+}
+
+// serveRequest creates the session for the incoming events and blocks the client.
+func (a *adapter) serveRequest(ctx context.Context, correlationID string, event cloudevents.Event) (*cloudevents.Event, cloudevents.Result) {
+	a.logger.Debugf("Handling request %q", correlationID)
+
+	respChan := a.sessions.add(correlationID)
+	defer a.sessions.delete(correlationID)
+
+	sendErr := make(chan error)
+	defer close(sendErr)
+
+	go func(errChan chan error) {
+		if res := a.ceClient.Send(cloudevents.ContextWithTarget(ctx, a.sinkURL), event); cloudevents.IsUndelivered(res) {
+			errChan <- res
+		}
+	}(sendErr)
+
+	a.logger.Debugf("Request forwarded to %q", a.sinkURL)
+
+	t := time.NewTimer(time.Duration(a.responseTimeout))
+	defer t.Stop()
```

Comment:
Maybe using `context.WithTimeout` would be a better fit here.

Reply:
I can't see how `context.WithTimeout` can be utilized here.